### PR TITLE
Continue previous run

### PR DESCRIPTION
**Pull request by nilin**
# Test

### pytest tests/integrations/train/test_runners.py -k 'test_reload_append' --run_very_slow

This runs an example for 10 epochs as run_1. It then reloads from the checkpoint at 5 epochs and re-runs the last epochs until 10 epochs total is reached again. The energies from the two runs are compared.

# Example

### Run for 110 epochs 
```
vmc-molecule \
  --presets.name=quicktest \
  --config.vmc.nepochs=110
```

### Discard epochs 101-110 and run epochs 101-200 as if it had been a single run of 200 epochs
```
vmc-molecule \
  --reload.logdir=logs/quicktest \
  --reload.same_logdir \
  --config.vmc.nepochs=200
```
Epochs are counted from the start of the reloaded run only if reload_config.append is True.

## Continue for 50 epochs with a new optimizer

To save in **logs/quicktest_proxsr**
```
vmc-molecule \
  --reload.logdir=logs/quicktest \
  --reload.new_optimizer \
  --config.vmc.optimizer_type=proxsr \
  --config.vmc.nepochs=250 \
  --config.subfolder_name=quicktest_proxsr
```

To save in **logs/proxsr/quicktest**
```
vmc-molecule \
  --reload.logdir=logs/quicktest \
  --reload.new_optimizer \
  --config.vmc.optimizer_type=proxsr \
  --config.vmc.nepochs=250 \
  --config.logdir=logs/proxsr
```

To avoid named subfolder behavior, do not define subfolder_name in the presets or flags

### CODE REVIEW

File: vmcnet/utils/io.py [hunk 1]
```diff
@@ -34,6 +34,26 @@ def append_metric_to_file(new_metric, logdir, name):
         np.savetxt(outfile, dumped_metric)
 
 
+def copy_txt_stats(source_dir, target_dir, truncate=None):
+    """Copies and truncated energy.txt as similar stats."""
+    names = ["energy", "energy_noclip", "variance", "variance_noclip", "accept_ratio"]
```
Comment (ggoldsh) [inline vmcnet/utils/io.py:39]: This is cool, the only part I don't like is that it's these 5 metrics only whereas there are options like `record_amplitudes` that result in other log files being made. I think we could handle this if we push the call to `copy_txt_stats` into the `vmc.py` `vmc_loop` function. Then we could do one iteration of the `update_param_fn`, get back a `metrics` object, and use the keys of the `metrics` dict to decide what data to copy over. It would probably add a little bit of extra threading of arguments around but it might be worth it to make the functionality more generic. What do you think?

File: vmcnet/train/runners.py [hunk 2]
```diff
@@ -635,29 +631,41 @@ def run_molecule() -> None:
...
+        (
+            reload_at_epoch,
+            data,
+            params,
+            reloaded_optimizer_state,
+            key,
+        ) = utils.io.reload_vmc_state(directory, filename)
+
+        if reload_config.append:
+            start_epoch = reload_at_epoch
```
Comment (ggoldsh) [inline vmcnet/train/runners.py:651]: One thing to note is that this change of `start_epoch` is actually a very important effect of the `append` flag, since this affects not just the logging but also the learning rate that is used. In other words, whether or not you want the text files to be concatenated, the only way to get the behavior of truly continuing a previous run, with the same learning rate schedule it would have had, is to use `reload_config.append`. I think that's probably fine but just wanted to point it out.

Comment (ggoldsh) [reply vmcnet/train/runners.py:651]: Idk maybe we should separate out those two pieces into different flags or rename this flag to something like `reload_config.continue_epoch_count.` Not totally sure.

File: vmcnet/train/default_config.py [hunk 3]
```diff
@@ -55,8 +55,11 @@ def get_default_reload_config() -> ConfigDict:
             "use_config_file": True,
             "config_relative_file_path": DEFAULT_CONFIG_FILE_NAME,
             "use_checkpoint_file": True,
-            "checkpoint_relative_file_path": CHECKPOINT_FILE_NAME,
-            "model_params_only": False,
+            "checkpoint_relative_file_path": RECENT_CHECKPOINT_FILE_NAME,
+            "new_optimizer": False,
```
[the remark this answers is not in the document]
Comment (nilin) [reply vmcnet/train/default_config.py]: renamed to new_optimizer_state to make it clear that it's about resetting the state

File: vmcnet/train/runners.py [hunk 4]
```diff
@@ -524,25 +524,28 @@ def _burn_and_run_vmc(
...
         checkpoint_dir = ""
         checkpoint_variance_scale = 0
         nhistory_max = 0
         check_for_nans = False
 
-    data, key = mcmc.metropolis.burn_data(
-        burning_step, run_config.nburn, params, data, key
-    )
+    if not skip_burn:
+        data, key = mcmc.metropolis.burn_data(
```
[the remark this answers is not in the document]
Comment (nilin) [reply vmcnet/train/runners.py:546]: this is meant for when a run is continued, so I want to avoid getting it confused with the original nburn when the run was started

File: vmcnet/train/runners.py (hunk 2 above)
Comment (nilin) [reply vmcnet/train/runners.py]: I've changed it to use the advanced start_epoch iff the old optimizer state is used

File: vmcnet/utils/io.py (hunk 1 above)
Comment (nilin) [reply vmcnet/utils/io.py:39]: changed to move all txt files except git_hash

File: vmcnet/utils/checkpoint.py [hunk 5]
```diff
@@ -608,6 +616,15 @@ def save_metrics_and_regular_checkpoint(
             )
             checkpoint_str = checkpoint_str + ", regular ckpt saved"
 
+    if recent_checkpoint_every is not None:
```
[the remark this answers is not in the document]
Comment (nilin) [reply vmcnet/utils/checkpoint.py]: outdated

File: vmcnet/train/runners.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ggoldsh) [reply vmcnet/train/runners.py:546]: Hm I guess so, I think I would prefer handling this in parse_config_flags by setting config.vmc.nburn=0 if reburn=False, and keeping this code oblivious to the particular reason that it is or is not burning. But if you prefer this way it's fine.

**Review by ggoldsh — APPROVED** [review]

**Comment by ggoldsh** [comment]
LGTM, thanks!
